EventProc/src/OverlayProducer.cxx

```diff
@@ -0,0 +1,329 @@
+		  
+		std::vector<SimCalorimeterHit> simHits = event.getCollection<SimCalorimeterHit>( collections_[iColl] );
+
+		// here: check if it exists -- if it's a simcalo hit collection. then define the out coll:
+
+		std::vector<SimCalorimeterHit> outHits;
+
+		bool isEcalHitCollection = false ;
+
+		if (strstr (collections_[iColl].c_str(), "Ecal") )
+		  isEcalHitCollection = true ;
+
+		std::vector<SimCalorimeterHit> overlayHits = overlayEvent->getCollection<SimCalorimeterHit>( collections_[iColl] );
+		//		std::vector<SimTrackerHit> overlayTrackerHits = overlayEvent->getCollection<SimTrackerHit>( collections_[iColl] );
+		  //dynamic_cast<SimCalorimeterHit> (overlayEvent->getCollection<SimCalorimeterHit>( collections_[iColl] )  );
+
+
+		//do something else if we're getting tracker hits. this type of check, for one, doesn't work, and simply leads to segfault if it's a simtracker hit
+		/*
+		if ( dynamic_cast<std::vector<SimCalorimeterHit>*>( &overlayHits  ) == NULL ) {
+		  
+		   ldmx_log(debug) <<  "Nope, no simcalorimeter hit collection called " << collections_[iColl] << ", skipping"  ;
+		  continue;
+		}
+		else
+		   ldmx_log(debug) <<  "It's simcalo alright" ;
+		overlayHits[0].Print();
+
+		*/
+		//if we alredy added at least one overlay event, this collection already exists 
+		if ( v_outHits.size() > iColl  ) {
+		  outHits = v_outHits[iColl];
+		} else { //otherwise, start out by just copying the sim hits, unaltered. 
+
+		  outHits=simHits;
+		  outCollNames.push_back( collections_[iColl]+"Overlay" );
+		  
+		  if (verbosity_ > 2) {
+			 ldmx_log(debug) <<  "[ OverlayProducer ]: in loop: start of collection " << outCollNames.back()  ;
+			 ldmx_log(debug) <<  "[ OverlayProducer ]: in loop: printing current sim event: "  ;
+		  }
+
+		  //we don't need to touch the hard process sim hits, really... but we might need the simhits in the hit map.
+		  for (const SimCalorimeterHit &simHit : simHits ) {
+			
+			simHit.Print();
+
+			//store ids of existing hits. for these, need to add contribs...
+			if (isEcalHitCollection) {
+			  int simHitID = simHit.getID();
+			  hitMap[ simHitID ] = SimCalorimeterHit();
+			  hitMap[ simHitID ].setID( simHitID );
+
+			  std::vector <float> hitPos = simHit.getPosition();
+			  hitMap[ simHitID ].setPosition( hitPos[0], hitPos[1], hitPos[2]);
+			}
+			
+		  } //over calo simhit collection
+		  
+		  
+		}
+		
+		if (verbosity_ > 2) {
+		   ldmx_log(debug) <<  "[ OverlayProducer ]: in loop: printing overlay event: "  ;
+		}
+
+		for ( SimCalorimeterHit &overlayHit : overlayHits ) {
+			
+		  overlayHit.Print();
+		  const float overlayTime = overlayHit.getTime() + timeOffset;
+		  overlayHit.setTime( overlayTime ); 
+
+		  if (! isEcalHitCollection )  //special treatment for ecal
+			outHits.push_back( overlayHit );
+		  else {
+			int overlayHitID = overlayHit.getID();
+			if ( hitMap.find( overlayHitID ) == hitMap.end() ) {
+			  hitMap[ overlayHitID ] = overlayHit;
```

Comment:
This line is what is causing the "doubling". If there isn't a regular sim hit in a cell, then we create a new `SimCalorimeterHit`. This line (in its current form) preserves all of the contribs from the overlay hit. Replacing this code block with something like the following should fix it.
```
hitMap[overlayHitID] = SimCalorimeterHit();
hitMap[ overlayHitID ].setID(overlayHitID);
std::vector <float> hitPos = overlayHit.getPosition();
hitMap[ overlayHitID ].setPosition( hitPos[0], hitPos[1], hitPos[2]);
hitMap[ overlayHitID ].addContrib(overlayIncidentID_,overlayTrackID_,overlayPdgCode_, overlayHit.getEdep(), overlayTime);
```

Reply:
Thanks – actually the duplication of simhits came from copying over the simhits collection in the start of the event, and then again when copying over all hits in the hitmap (after adding in overlay contribs). I've fixed this now. But I did take this suggestion (basically it's the inverse of what I did before) since it makes sure we're not preserving any overlay hit contribs.